### PR TITLE
Add BFloat16 support and optimization for mish, hardtanh backward, and silu on CPU

### PR DESCRIPTION
**Pull request by CaoE**
### Description
* add BFloat16 support for mish and hardtanh backward on CPU.
* optimize the performance for silu

### Testing

- optimize the performance for silu: bfloat16

single socket (28 cores):
```
before: 1x128x1024  forward 0.090 s  backward  0.218 s
        10x128x1024 forward 0.146 s  backward  0.314 s
            
after:  1x128x1024   forward  0.064 s backward  0.100 s
        10x128x1024  forward  0.085 s backward  0.133 s
```
single core:
```
before: 1x128x1024   forward 0.300 s  backward  0.606 s
        10x128x1024  forward 2.825 s  backward  5.834 s

after:  1x128x1024   forward 0.156 s backward   0.239 s
        10x128x1024  forward 1.447 s backward   2.165 s
```

- Add BFloat16 support for mish and backward of hardtanh on CPU.

single socket (20 cores):
op | shape | fp32 / s | fp32 / s | bf16 / s |  bf16 / s
-- | -- | -- | -- | -- | --
  |   | forward | backward | forward | backward
silu | [10, 128, 10, 10] | 4.41E-05 | 7.67E-05 | 5.32E-05 | 9.38E-05
  | [10, 128, 80, 80] | 0.0008 | 0.001788 | 0.00067 | 0.001031
mish | [10, 128, 10, 10] | 0.000356 | 0.000427 | 0.000367 | 0.000436
  | [10, 128, 80, 80] | 0.004527 | 0.005807 | 0.004757 | 0.005393
hardtanh | [10, 128, 10, 10] | / | 3.97E-05 | / | 4.45E-05
  | [10, 128, 80, 80] | / | 0.001748 | / | 0.000645

single core:
op | shape | fp32 / s | fp32 / s | bf16 / s |  bf16 / s
-- | -- | -- | -- | -- | --
  |   | forward | backward | forward | backward
silu | [10, 128, 10, 10] | 1.17E-04 | 1.91E-04 | 1.35E-04 | 2.23E-04
  | [10, 128, 80, 80] | 0.007434 | 0.013141 | 0.008464 | 0.013044
mish | [10, 128, 10, 10] | 0.00103 | 0.00122 | 0.00106 | 0.001227
  | [10, 128, 80, 80] | 0.065629 | 0.078418 | 0.067779 | 0.077214
hardtanh | [10, 128, 10, 10] | / | 1.18E-04 | / | 9.30E-05
  | [10, 128, 80, 80] | / | 0.010773 | / | 0.005834

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by CaoE** [comment]
Hi @frank-wei , could you please review this PR ? Thank you.

**Comment by jgong5** [comment]
@CaoE Do you mind make the PR title more specific about what changes you made? The word "activation" sounds too general.

**Comment by CaoE** [comment]
Changed the title, and I will provide more performance numbers later.

**Review by mingfeima — APPROVED** [review]

**Comment by CaoE** [comment]
Hi @kit1980, could you please view this PR ? Thank you.

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 1]
```diff
@@ -623,7 +623,25 @@ void shrink_backward_kernel(TensorIteratorBase& iter, const Scalar& lambd) {
...
 
 void hardtanh_backward_kernel(TensorIterator& iter, const Scalar& min, const Scalar& max) {
-  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "hardshrink_backward_cpu", [&] {
+  if (iter.dtype() == kBFloat16) {
+    auto min_val = min.to<float>();
+    auto max_val = max.to<float>();
+    cpu_kernel_vec(
+        iter,
+        [=](BFloat16 grad_val, BFloat16 self_val) -> BFloat16 {
+          return (float(self_val) <= min_val || float(self_val) >= max_val) ? float(0) : float(grad_val);
```
Comment (malfet) [inline aten/src/ATen/native/cpu/Activation.cpp:632]: Nit
```suggestion
          return (float(self_val) <= min_val || float(self_val) >= max_val) ? BFloat16(0) :grad_val;
```

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 2]
```diff
@@ -1035,8 +1054,23 @@ void glu_backward_kernel(TensorIterator& iter) {
 }
 
 void silu_kernel(TensorIteratorBase& iter) {
-  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(
-      kBFloat16, iter.dtype(), "silu_cpu", [&]() {
+  if (iter.dtype() == kBFloat16) {
+      const Vectorized<float> kOneVec(float(1));
```
Comment (malfet) [inline aten/src/ATen/native/cpu/Activation.cpp:1058]: Nit
```suggestion
      const Vectorized<float> kOneVec(1.0f);
```

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 3]
```diff
@@ -1035,8 +1054,23 @@ void glu_backward_kernel(TensorIterator& iter) {
...
+      cpu_kernel_vec(
+          iter,
+          [](BFloat16 x) -> BFloat16 {
+            return float(x) / (float(1) + std::exp(-float(x)));
```
Comment (malfet) [inline aten/src/ATen/native/cpu/Activation.cpp:1062]: Nit
```suggestion
            return float(x) / (1.0f + std::exp(-float(x)));
```

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 4]
```diff
@@ -1047,11 +1081,34 @@ void silu_kernel(TensorIteratorBase& iter) {
...
             });
       });
+    }
 }
 
 void silu_backward_kernel(TensorIteratorBase& iter) {
-  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(
-      kBFloat16, iter.dtype(), "silu_backward_cpu", [&]() {
+  if (iter.dtype() == kBFloat16) {
+    const Vectorized<float> kOneVec(float(1));
```
Comment (malfet) [inline aten/src/ATen/native/cpu/Activation.cpp:1089]: Nit
```suggestion
    const Vectorized<float> kOneVec(1.0f);
```

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 5]
```diff
@@ -1047,11 +1081,34 @@ void silu_kernel(TensorIteratorBase& iter) {
...
+    cpu_kernel_vec(
+        iter,
+        [](BFloat16 dy, BFloat16 x) -> BFloat16 {
+          const float sigmoid =
+              float(1) / (float(1) + std::exp(-float(x)));
+          return dy * sigmoid * (float(1) + x * (float(1) - sigmoid));
```
Comment (malfet) [inline aten/src/ATen/native/cpu/Activation.cpp:1095]:
```suggestion
              1.0f / (1.0f + std::exp(-float(x)));
          return dy * sigmoid * (1.0f + x * (1.0f - sigmoid));
```

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 6]
```diff
@@ -1080,10 +1153,36 @@ void mish_kernel(TensorIteratorBase& iter) {
...
             });
       });
+  }
 }
 
 void mish_backward_kernel(TensorIterator& iter) {
-  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "mish_backward_cpu", [&]() {
+  if (iter.dtype() == kBFloat16) {
+    using Vec = Vectorized<float>;
+    const Vec kOneVec(float(1));
```
Comment (malfet) [inline aten/src/ATen/native/cpu/Activation.cpp:1162]:
```suggestion
    const Vec kOneVec(1.0f);
```

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 7]
```diff
@@ -1080,10 +1153,36 @@ void mish_kernel(TensorIteratorBase& iter) {
...
+    cpu_kernel_vec(
+        iter,
+        [](BFloat16 dy, BFloat16 x) -> BFloat16 {
+          const float sigmoid =
+              float(1) / (float(1) + std::exp(-float(x)));
```
Comment (malfet) [inline aten/src/ATen/native/cpu/Activation.cpp:1167]:
```suggestion
              1.0f / (1.0f + std::exp(-float(x)));
```

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 8]
```diff
@@ -1080,10 +1153,36 @@ void mish_kernel(TensorIteratorBase& iter) {
...
+          const float tanh_softplus = std::tanh(std::log1p(std::exp(float(x))));
+          return dy * (tanh_softplus + x * sigmoid * (float(1) - tanh_softplus * tanh_softplus));
```
Comment (malfet) [inline aten/src/ATen/native/cpu/Activation.cpp:1169]:
```suggestion
          return dy * (tanh_softplus + x * sigmoid * (1.0f - tanh_softplus * tanh_softplus));
```

**Review by malfet — APPROVED** [review]
LGTM, but see nits

**Pull request closed without merging**